documentation/directory-structure.md

```diff
@@ -30,13 +34,66 @@ You may be also interested in following:
 > These exceptions are not allowed going forward.
 
+## Key concepts
+
+The directory structure is a reflection of few key concepts.
+
+A `service` is a set of operation endpoints (typically HTTP REST API endpoints) that **version uniformly**.
+
+A `service group` is a set of services that share common (ARM) **Resource Provider (RP) Namespace**, `RPNS`.
+
+In case of ARM, a `service` consists of multiple `resource types`.
+
+### Uniform versioning
+
+A `service` **must version uniformly**. This means that the service,
+all its dependencies, and all artifacts generated from the service,
+**must** have the same version.
+
+In practice, this means, for an example service with version `2024-03-25`:
+
+- All the OpenAPI specification `.json` files the service is composed of must have the same version of `2024-03-25`.
+- All the shared specifications (i.e. `common-types`) the service depends on must have the same version of `2024-03-25`.
+- In case of ARM service, all the resource types must have the same version of `2024-03-25`.
+- The language SDKs generated from the service must have the same version of `2024-03-25`.
+- The documentation published for the service must have the same version of `2024-03-25`.
+- Update to any version of the above requires update of all the other versions.
+  For example, if `common-types` published an updated version of `2024-04-17`,
+  then if the service wants to take dependency on it, it can only do it in
+  a new version of `2024-04-17`.
+  Because the service version is now `2024-04-17`, all the OpenAPI specification `.json` files the service is composed
+  of must have the same version of `2024-04-17`.
+  In addition, a new SDK must be generated from the service, and new documentation published, both tagged with
+  version `2024-04-17`.
+- As a consequence of the above, if given service version uses `-preview` version anywhere, then the service
+  version itself must be `-preview`, it can depend only **only** `-preview` versions, and the generated SDK
+  and published docs must denote they are in `preview` too.
+
+TODO: 
+- These rules seem too strict. What if cadence of releasing `common-types` does not align with cadence of releasing
+  given service? Same for resource types? 
+  - For example, service got released on `2024-01-01`, then new `common-types` got released on `2024-02-01` and on
+    `2024-03-01` new service version is to be released? Can a `2024-03-01` service version depend on `2024-02-01` `common-types`?
+  - If a service version can depend on `common-types` of a different version, can the generated SDK also depend
+    on different `.json` versions?
+    See e.g. [this bad setup of compute rm 2024-03-01](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/readme.md#tag-package-2024-03-01).
+    Could we make package `2024-03-01` depend on, say, `common-types` of `2024-02-01` and all other `.json` files of, say, `2024-02-15` ? 
```

Comment:
I think most of these concerns are moot if you accept the change I proposed above.